### PR TITLE
[AArch64] Fix heuristics for folding "lsl" into load/store ops.

### PR DESCRIPTION
**Pull request by efriedma-quic**
The existing heuristics were assuming that every core behaves like an Apple A7, where any extend/shift costs an extra micro-op... but in reality, nothing else behaves like that.

On some older Cortex designs, shifts by 1 or 4 cost extra, but all other shifts/extensions are free.  On all other cores, as far as I can tell, all shifts/extensions for integer loads are free (i.e. the same cost as an unshifted load).

To reflect this, this patch:

- Enables aggressive folding of shifts into loads by default.

- Removes the old AddrLSLFast feature, since it applies to everything except A7 (and even if you are explicitly targeting A7, we want to assume extensions are free because the code will almost always run on a newer core).

- Adds a new feature AddrLSLSlow14 that applies specifically to the Cortex cores where shifts by 1 or 4 cost extra.

I didn't add support for AddrLSLSlow14 on the GlobalISel side because it would require a bunch of refactoring to work correctly. Someone can pick this up as a followup.

### CODE REVIEW

**Comment by efriedma-quic** [comment]
I looked at some of the older reviews, and rechecked my research... a few more notes:

- It turns out the current commit message isn't precisely right.  Current Cortex cores (X2 and later) have free shift by 1 for integer loads... but no free shift by 1/4 for floating-point loads.  Not sure if it's worth explicitly modeling int-shift vs. float-shift.
- https://reviews.llvm.org/D155470#4527270 suggests that we should default to AddrLSLSlow14... I'm not sure if that's the right choice. An explicit shift is guaranteed to increase latency, but an extra integer micro-op generated by a folded shift might not matter in a lot of cases.

File: llvm/lib/Target/AArch64/AArch64.td [hunk 1]
```diff
@@ -1268,7 +1270,6 @@ def TuneNeoverseN1 : SubtargetFeature<"neoversen1", "ARMProcFamily", "NeoverseN1
                                       "Neoverse N1 ARM processors", [
                                       FeatureFuseAES,
                                       FeatureFuseAdrpAdd,
-                                      FeatureAddrLSLFast,
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64.td:1271]: +FeatureAddrLSLSlow14 here I would expect?

File: llvm/lib/Target/AArch64/AArch64.td [hunk 2]
```diff
@@ -1298,7 +1297,6 @@ def TuneNeoverseV1 : SubtargetFeature<"neoversev1", "ARMProcFamily", "NeoverseV1
                                       "Neoverse V1 ARM processors", [
                                       FeatureFuseAES,
                                       FeatureFuseAdrpAdd,
-                                      FeatureAddrLSLFast,
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64.td:1301]: +FeatureAddrLSLSlow14

**Review by davemgreen — COMMENTED** [review]
Thanks for looking into this, it is one thing off my todo list.

> It turns out the current commit message isn't precisely right. Current Cortex cores (X2 and later) have free shift by 1 for integer loads... but no free shift by 1/4 for floating-point loads. Not sure if it's worth explicitly modeling int-shift vs. float-shift.

That might be an inaccuracy in the optimization guide more than a difference between int/fp. lsl #4 are still an extra operation, but I do not believe they come up very often.

> https://reviews.llvm.org/D155470#4527270 suggests that we should default to AddrLSLSlow14... I'm not sure if that's the right choice. An explicit shift is guaranteed to increase latency, but an extra integer micro-op generated by a folded shift might not matter in a lot of cases.

Yeah I agree - with more new cores having faster #1 shifts I think it should be OK as the default.

File: llvm/lib/Target/AArch64/AArch64.td (hunk 1 above)
Comment (efriedma-quic) [reply llvm/lib/Target/AArch64/AArch64.td:1271]: Done

File: llvm/lib/Target/AArch64/AArch64.td (hunk 2 above)
Comment (efriedma-quic) [reply llvm/lib/Target/AArch64/AArch64.td:1301]: Done

**Review by davemgreen — APPROVED** [review]
LGMT, so long as the tests are OK. Thanks

**Comment by AtariDreams** [comment]
Tests did not go well, though it seems to be no fault of this PR. Rebase and try again @efriedma-quic

**Review by fhahn — APPROVED** [review]
LGTM for the Apple cores, thanks!

**Pull request merged**